### ISSUE TITLE
Fix binary (un-)pickling of PyVista objects

### ISSUE THREAD

**Issue by Onandon11** [pyvista/pyvista#1768]
Some time ago support was added for pickling/unplickling of PyVista dataobject. This feature is really useful in case of inter process communication like multi processing. However the feature is implemented use a ascii writer, which is tremendously slow.

I played somewhat with the code of @banesullivan (see: https://github.com/pyvista/pyvista-support/issues/59#issue-494762368) and tried to use the binary writer. I found out it was due to the `pickle.dump` the arrays were getting messed up. If you don't pipe it to an output string but to a file the example works like a charm.

Of course some functionality has be added to removed the files after they are red but I would really like to see the `SetFileTypeToBinary` to be the default.

```python
import vtk
import pickle
import pyvista as pv
import numpy as np

def pickle_vtk(mesh, filename):
    writer = vtk.vtkDataSetWriter()
    writer.SetInputDataObject(mesh)
    writer.SetFileName(filename)
    writer.SetFileTypeToBinary()
    writer.Write()
    return filename


def unpickle_vtk(filename):
    reader = vtk.vtkDataSetReader()
    reader.SetFileName(filename)
    reader.Update()
    return pv.wrap(reader.GetOutput())


#### Example

from pyvista import examples
mesh = examples.load_random_hills()

fname = 'filename.vtkpickle'
pickle_vtk(mesh, fname)
test = unpickle_vtk(fname)

assert np.allclose(mesh.points, test.points)
assert mesh.n_arrays == test.n_arrays
```

**Comment by MatthewFlamm** [comment]
We don't even need to be able to cleanup the file if you use a BytesIO object, which I think should be pickle-able.

```py
import vtk
import pickle
import pyvista as pv
import numpy as np
import io

def pickle_vtk(mesh, filename):
    writer = vtk.vtkDataSetWriter()
    writer.SetInputDataObject(mesh)
    writer.SetFileName(filename)
    writer.SetFileTypeToBinary()
    writer.Write()
    return filename


def unpickle_vtk(filename):
    reader = vtk.vtkDataSetReader()
    reader.SetFileName(filename)
    reader.Update()
    return pv.wrap(reader.GetOutput())


#### Example

from pyvista import examples
mesh = examples.load_random_hills()

fname = io.BytesIO()
pickle_vtk(mesh, fname)
test = unpickle_vtk(fname)

assert np.allclose(mesh.points, test.points)
assert mesh.n_arrays == test.n_arrays
```

**Comment by whophil** [comment]
For reference https://github.com/pyvista/pyvista/pull/1143/

**Comment by Onandon11** [comment]
If I look in the referenced PR, the binary mode isn't working. Maybe it's fixed in the latest vtk version 9? 

The suggestion of using a BytesIO object is great, I forgot about that one!

**Comment by whophil** [comment]
@Onandon11 please take a look at #1778 and see if this resolves your issue.

Unfortunately, BytesIO didn't work here because VTK methods which take filenames want real strings.

**Comment by MatthewFlamm** [comment]
> Unfortunately, BytesIO didn't work here because VTK methods which take filenames want real strings.

It works in my testing.

**Comment by adam-grant-hendry** [comment]
@akaszynski @banesullivan @adeak @tkoyama010 @MatthewFlamm This seems to break depending on dataset size. e.g.

```python
import pickle

import numpy as np
import pyvista as pv
from vtkmodules.vtkIOLegacy import vtkDataSetWriter


def pickle_vtk(mesh, filename):
    writer = vtkDataSetWriter()
    writer.SetInputDataObject(mesh)
    writer.SetWriteToOutputString(True)
    writer.SetFileTypeToBinary()
    writer.Write()
    to_serialize = writer.GetOutputString()

    with open(filename, 'wb') as handle:
        pickle.dump(to_serialize, handle, protocol=pickle.HIGHEST_PROTOCOL)

    return filename


if __name__ == '__main__':
    dims = (2154, 1500, 1167)

    volume = pv.UniformGrid(
        dims=(2154, 1500, 1167),
        spacing=(1, 1, 1),
        origin=(0, 0, 0),
    )

    volume.point_data['scalars'] = np.zeros(
        shape=(dims[0] * dims[1] * dims[2],), dtype=np.uint8
    )

    fname = 'filename.vtkpickle'
    pickle_vtk(volume, fname)  # Crashes
```

However, I have no problem pickling the numpy array by itself to file (this is 3.7 GB).

**Comment by MatthewFlamm** [comment]
Have you tried the more modern xml dataset writer?

**Comment by adeak** [comment]
@adam-grant-hendry it's generally more helpful to be more specific than "breaks"/"crashes". How does it break? How does it crash? Is there a segfault? An exception? Do you have a traceback? These are key details if you post the same thing in multiple places, pinging the same 5 people.

I'd see for myself but this snippet cut through my free 8 gigs of RAM, taking out my laptop. Technically crashing but I doubt this is what you had in mind.

**Comment by whophil** [comment]
> Have you tried the more modern xml dataset writer?

I would also try this - I believe @MatthewFlamm is referring to https://vtk.org/doc/nightly/html/classvtkDataSetWriter.html. 

There is (or perhaps _was_) a bug which made it impossible to use `vtkXMLDataSetWriter` in the binary pickling code when it was implemented - see https://github.com/pyvista/pyvista/pull/1778#issuecomment-960433089. Specifically, that writer was not able to write to string. However, it should be able to write to file, and perhaps it works better for large files.

**Comment by adam-grant-hendry** [comment]
> @adam-grant-hendry it's generally more helpful to be more specific than "breaks"/"crashes

Not to be rude, truly, but this ain't my first rodeo. Also, unlike the average Joe, I do contribute PRs.

Regardless, this is about as bad as it gets. There is no segfault, exception, or traceback. It simply stops running. Consequently, I do believe this is a problem on VTK's end.

**Comment by adam-grant-hendry** [comment]
@adeak You're also welcome to try the example I provided. You will see there is very little to go off of with the end result.

**Comment by MatthewFlamm** [comment]
I think @adeak comment could have separated better between the ask for more information and asking for better consideration of the people being pinged. This is my interpretation of the comment. Although I agree with both points separately.

The pinging seems nonspecific here and misses key contributors to this functionality.  Large amounts of pings should be reserved for when there is a critical bug or a large blocker for continued development.  This particular issue is localized.

Having two places for this conversation will be distracting, can we move this discussion to #3269 so that it is easier to track?

**Comment by adam-grant-hendry** [comment]
> @adeak You're also welcome to try the example I provided. You will see there is very little to go off of with the end result.

I should have read the rest of your comment that trying to test would have crashed your laptop! My bad!

**Comment by adam-grant-hendry** [comment]
> > Have you tried the more modern xml dataset writer?
> 
> I would also try this - I believe @MatthewFlamm is referring to https://vtk.org/doc/nightly/html/classvtkDataSetWriter.html.
> 
> There is (or perhaps _was_) a bug which made it impossible to use `vtkXMLDataSetWriter` in the binary pickling code when it was implemented - see [#1778 (comment)](https://github.com/pyvista/pyvista/pull/1778#issuecomment-960433089). Specifically, that writer was not able to write to string. However, it should be able to write to file, and perhaps it works better for large files.

Thank you for this suggestion. I will try it out if I have time.

At the moment, I am interested in pickling for serialization/marshaling (similar to that mentioned by the [OP](https://github.com/pyvista/pyvista/issues/1768#issue-1041129686)), but I will take a look.

I find it interesting that I can pull the `numpy` points dataset and pickle that to file without issue, but that [`writer.Write()` causes the program to immediately end](https://github.com/pyvista/pyvista/issues/3269#issuecomment-1237366724). If you can offer any assistance in this area (over at #3269 ), I would greatly appreciate it!

**Comment by whophil** [comment]
> There is (or perhaps _was_) a bug which made it impossible to use `vtkXMLDataSetWriter` in the binary pickling code when it was implemented - see [#1778 (comment)](https://github.com/pyvista/pyvista/pull/1778#issuecomment-960433089). Specifically, that writer was not able to write to string. However, it should be able to write to file, and perhaps it works better for large files.

To respond to myself - I looked into this a bit, and it seems like it would be a one-line change in VTK to get `vtkXMLDataSetWriter` working for our use cases. I opened this issue on the VTK tracker https://gitlab.kitware.com/vtk/vtk/-/issues/18661. Even with the fix merged today, it would likely be some time before this change made its way into the official VTK wheels and non-official conda packages.
